### DISCUSSION TITLE
More acturate touchpad horizontal swipe gesture

### DISCUSSION THREAD

**Discussion by lost-melody** [YaLTeR/niri#832] [Ideas]
I've been using niri for a while and have noticed that I cannot easily focus a window by swiping.

If I have three windows in a workspace and want to swipe from the first one to the second one, I have to swipe very carefully and avoid swiping too fast or too much distance, or I would skip over the second one and eventually focus the third one.

The swipe gesture now moves windows responsively, like what paperwm does, but there's no indicators telling me which window will be focused on fingers released. And windows don't stay where they are when releasing fingers, instead they snap to screen edges.

For this reason, I don't think it necessary to responsively move windows during swiping, we only need to move the "focus ring" to the correct window, indicating that the window would be focused, like what hyprscroller does.

Or if we prefer moving windows responsively during swiping, we'd better also responsively indicate the focus ring to the correct window, and let windows stay where they are when swiping ends.

**Comment by YaLTeR** [comment]
I definitely prefer the gestures to cause responsive movement in all cases. That's like the whole point of having 1-to-1 gestures.

I understand the problem you're having, but I'm not sure a focus ring indicator will work too well because the final window is selected only when you release the fingers, and takes into account the velocity at the point of release. The horizontal swipe gesture wasn't really meant for super precise window picking.

**Reply by lost-melody** [reply]
I see. I guess the velocity now affects too much, especially the moment when I release my fingers. Even if I slow down my fingers and want to focus on the current window, the tiny movement in the moment I release my fingers would cause a focus change. Possible to tweak the velocity with a lower factor? Or filter out those tiny movements?

**Reply by YaLTeR** [reply]
Also after taking into account the velocity, it finds the closest window-monitor edge snap, and then it focuses the furthest window into the direction of the swipe. So you can't focus a window in the middle of the screen with a swipe.

**Reply by YaLTeR** [reply]
The idea was that with a swipe gesture you want to focus something partially or fully off screen.

**Reply by lost-melody** [reply]
I see, thank you.
I'm more used to move a window center to focus on it, and that's why I'm confused about the swipe gesture.

**Reply by lost-melody** [reply]
After 3 days using, I now understand the "furthest window" mechanism, so can accurately select windows, but I still can't get used to it.
If all of my windows are full-width, it is great to have this mechanism. But I prefer my windows to be of half screen width, so if I want to focus on the window next to the current window, I have to select it by clicking on it, or I have to swipe reversely first then backward until the target window is half off screen.
So I think the "off-screen window" should be selected only when there's no window fully in view, and if I actually want to select the window off screen, I can always swipe a little more to make it fully in view, and that's what I think a more intuitive way.

**Reply by YaLTeR** [reply]
> I have to select it by clicking on it

What about [focus-follows-mouse](https://github.com/YaLTeR/niri/wiki/Configuration:-Input#focus-follows-mouse)?

```
input {
    // Allow focus-follows-mouse only when it will not scroll the view.
    focus-follows-mouse max-scroll-amount="0%"
}
```

**Reply by lost-melody** [reply]
Thank you, I will try it for a while. I've never used focus-follows-mouse before, are there any side effects?

Edit:
Good. But there's a similar situation (to the swiping reversely and back): if my mouse is currently in the target window while focus is on another window (focus previously moved by keybindings), I will need to move cursor out and then in to focus on it.
Edit:
In this case, a single click will work :)

**Reply by YaLTeR** [reply]
With `max-scroll-amount="0%"`, not really. Give it a try.

**Reply by mastoca** [reply]
is there a way to completely disable snapping? 

I have a dedicated trackpad to just be responsible for sliding around and would prefer snapping via a key binding. This would give the feel that the canvas is mine to control instead of it trying to snap to the closest.

the `max-scroll-amount="0%"` helps tremendously...but the windows still fidget too much which is meh. For example, sometimes when neighboring window widths are huge (like two panels at >70% of whole width of screen) this gets quite jumpy.

**Reply by YaLTeR** [reply]
Not at the moment. Also, some snapping will always occur because the focused window is always scrolled fully into view.

**Reply by mastoca** [reply]
yes, it sounds like that's what I was trying to disable, the requirement that the focused window is scrolled fully into view. I understand it is a weird case. If there was a percent that a focused window **has** to come into view...I would likely set it around 15-20% so I can see the cursor and what I'm typing without the 'jitteryness'. 

eg. repeating a command in a terminal (which I don't need to see) while watching the results on the current window I'm interested in.

**Reply by YaLTeR** [reply]
Focused window scrolling fully into view is pretty thoroughly baked into the code, I'm afraid there's no simple way to tweak it.

**Comment by lost-melody** [comment]
After some days using, I'm back now to talk about my feelings:

`focus-follows-mouse` feels good! And I prefer a value of about 5%~10%.
Larger horizontal margin (by configuring `struts`) also helps, so that I can click on windows right next to the current viewport, and it's fancy!
It's still not easy to focus on windows far away, and I often miss the target window if I swipe a little bit faster (too sensitive?). And it feels weird when I release my fingers carefully while the windows start to move (significantly) to snap to the screen edges.
Also, it is hard for me to get used to the mechanism of "focus the window partially off screen". :)

Anyway, with "focus follows mouse" it is now much easier to use. And for those "swipe" related problems, it's actually not that often for me to select windows far away so it doesn't bother me a lot.
Plus I wonder if there are ideas to center the current window by touchpad/mouse? Current I do this by left-click the "waybar-niri-window" widget to run a niri-action command, but that would need a mouse moving.

Thank you again!
Is it ready to close this discussion?

**Reply by YaLTeR** [reply]
Not sure if adding centered window positions to the horizontal gesture snapping is worth it. Maybe? Or would it get too annoying when you do want to snap to screen edges?

Feel free to close discussions as you see fit

**Reply by lost-melody** [reply]
I don't think it will worth it much, since it might end up to be a lot of work: to implement a switchable center/snap/free position mode option or keybinding.

But for focus moving, I believe it's necessary to also include centered windows because it is related to and can be tweaked by the "final velocity": If I do want to switch to off-screen windows, I will release with a larger velocity or simply swipe a little more distance; and if I mean to focus on windows centered in screen, I will release my fingers with a velocity of nearly zero.

And, yeah, it's not a very important issue so I think we'd put some other things first.
Closing this, thank you for your time! :)
